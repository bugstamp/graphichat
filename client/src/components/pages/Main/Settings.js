import React, {
  PureComponent,
  createRef,
} from 'react';
import PropTypes from 'prop-types';
import { map, omit } from 'lodash';

import TextField from '@material-ui/core/TextField';

import {
  SettingWrapper,
  AvatarWrapper,
  Avatar,
  Camera,
  FormWrapper,
} from './styled';

import { getAvatar } from '../../../helpers';
import { meProps } from '../../propTypes';

class Settings extends PureComponent {
  avatarInput = createRef()

  state = {
    errors: {},
    values: {},
    called: {},
  }

  fields = [
    {
      id: 'username',
      label: 'User Name',
    },
    {
      id: 'firstName',
      label: 'First Name',
    },
    {
      id: 'lastName',
      label: 'Last Name',
    },
  ]

  getSnapshotBeforeUpdate(prevProps) {
    const { error } = this.props;
    const snapshot = {
      error: null,
    };

    if (!prevProps.error && error) {
      const { graphQLErrors } = error;
      const { data = null } = graphQLErrors[0];

      if (data) {
        const { message } = graphQLErrors[0];
        const { invalidField = null } = data;

        if (invalidField) {
          snapshot.error = { [invalidField]: message };
        }
      }
    }
    return snapshot;
  }

  componentDidUpdate(prevProps, prevState, { error }) {
    if (error) {
      this.addError(error);
    }
  }

  onChange = ({ target }) => {
    const { name, value } = target;

    this.setState(({ called, values }) => {
      const updatedCalled = !called[name]
        ? { ...called, [name]: true }
        : called;
      const updatedValues = { ...values, [name]: value };

      return {
        called: updatedCalled,
        values: updatedValues,
      };
    });
  }

  onChangeAvatar = (e) => {
    const {
      target: {
        files: [file],
        validity: { valid },
      },
    } = e;
    const { uploadAvatar } = this.props;

    if (valid) {
      uploadAvatar({ variables: { file } });
    }
  }

  onAvatarClick = () => {
    if (this.avatarInput) {
      this.avatarInput.current.click();
    }
  }

  onSubmit = ({ target }) => {
    const { errors } = this.state;
    const { updateUser, me } = this.props;
    const { name: field, value } = target;
    const existedValue = me[field];

    if (value !== existedValue && value) {
      updateUser({ variables: { field, value } });

      if (errors[field]) {
        this.setState(() => ({
          errors: omit(errors, field),
        }));
      }
    } else if (!value) {
      this.onChange({ target: { name: field, value: existedValue } });
    }
  }

  onKeyPressed = (e) => {
    const { key, target } = e;

    if (key === 'Enter') {
      target.blur();
    }
  }

  addError = (error) => {
    this.setState(({ errors }) => ({
      errors: { ...errors, ...error },
    }));
  }

  render() {
    const { called, values, errors } = this.state;
    const { me } = this.props;
    const avatar = getAvatar(me, 'md');

    return (
      <SettingWrapper>
        <AvatarWrapper>
          <input
            ref={this.avatarInput}
            type="file"
            accept="image/png, image/jpg, image/jpeg"
            onChange={this.onChangeAvatar}
            style={{ display: 'none' }}
          />
          <Avatar
            src={avatar.src}
            alt={avatar.text}
            onClick={this.onAvatarClick}
          >
            <Camera fontSize="large" />
          </Avatar>
        </AvatarWrapper>
        <FormWrapper>
          {
            map(this.fields, ({ id, label }) => {
              const value = called[id] ? values[id] : me[id];
              const error = !!errors[id];
              const message = errors[id];

              return (
                <TextField
                  key={id}
                  name={id}
                  value={value}
                  label={label}
                  error={error}
                  helperText={message}
                  variant="outlined"
                  margin="dense"
                  onChange={this.onChange}
                  onBlur={this.onSubmit}
                  onKeyDown={this.onKeyPressed}
                  fullWidth
                  required
                />
              );
            })
          }
        </FormWrapper>
      </SettingWrapper>
    );
  }
}

Settings.defaultProps = {
  me: {},
  error: null,
};
Settings.propTypes = {
  me: PropTypes.shape(meProps),
  uploadAvatar: PropTypes.func.isRequired,
  updateUser: PropTypes.func.isRequired,
  error: PropTypes.instanceOf(Error),
};

export default Settings;
