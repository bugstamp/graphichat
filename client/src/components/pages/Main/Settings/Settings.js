import React, {
  PureComponent,
  createRef,
} from 'react';
import PropTypes from 'prop-types';
import { map, omit, isEqual } from 'lodash';

import TextField from '@material-ui/core/TextField';
import CheckIcon from '@material-ui/icons/CheckRounded';

import {
  SettingsWrapper,
  AvatarWrapper,
  Avatar,
  Camera,
  FormWrapper,
  InputCheckIconWrapper,
} from './styled';

import { getAvatar } from '../../../../helpers';
import { meProps } from '../../../propTypes';

class Settings extends PureComponent {
  avatarInput = createRef()

  state = {
    errors: {},
    values: {},
    called: {},
    success: '',
  }

  fields = [
    {
      id: 'username',
      label: 'User Name',
    },
    {
      id: 'firstName',
      label: 'First Name',
    },
    {
      id: 'lastName',
      label: 'Last Name',
    },
  ]

  getSnapshotBeforeUpdate(prevProps) {
    const { error, updateUserResult } = this.props;
    const snapshot = {
      error: null,
      success: null,
    };

    if (!prevProps.error && error) {
      const { graphQLErrors } = error;
      const { data = null } = graphQLErrors[0];

      if (data) {
        const { message } = graphQLErrors[0];
        const { invalidField = null } = data;

        if (invalidField) {
          snapshot.error = { [invalidField]: message };
        }
      }
    }
    if (!isEqual(prevProps.updateUserResult, updateUserResult)) {
      snapshot.success = updateUserResult.field || null;
    }
    return snapshot;
  }

  componentDidUpdate(prevProps, prevState, { error, success }) {
    if (error) {
      this.addError(error);
    }
    if (success) {
      this.setSuccess(success);
    }
  }

  setSuccess = success => this.setState({ success });

  onChange = ({ target }) => {
    const { name, value } = target;

    this.setState(({ called, values }) => {
      const updatedCalled = !called[name]
        ? { ...called, [name]: true }
        : called;
      const updatedValues = { ...values, [name]: value };

      return {
        called: updatedCalled,
        values: updatedValues,
      };
    });
  }

  onChangeAvatar = (e) => {
    const {
      target: {
        files: [file],
        validity: { valid },
      },
    } = e;
    const { uploadAvatar } = this.props;

    if (valid) {
      uploadAvatar({ variables: { file } });
    }
  }

  onAvatarClick = () => {
    if (this.avatarInput) {
      this.avatarInput.current.click();
    }
  }

  onSubmit = ({ target }) => {
    const { errors } = this.state;
    const { updateUser, me } = this.props;
    const { name: field, value } = target;
    const existedValue = me[field];

    if (value !== existedValue && value) {
      updateUser({ variables: { field, value } });

      if (errors[field]) {
        this.setState(() => ({
          errors: omit(errors, field),
        }));
      }
    } else if (!value) {
      this.onChange({ target: { name: field, value: existedValue } });
    }
  }

  onKeyPressed = (e) => {
    const { key, target } = e;

    if (key === 'Enter') {
      target.blur();
    }
  }

  addError = (error) => {
    this.setState(({ errors }) => ({
      errors: { ...errors, ...error },
    }));
  }

  render() {
    const {
      called,
      values,
      errors,
      success,
    } = this.state;
    const { me } = this.props;
    const avatar = getAvatar(me, 'md');

    return (
      <SettingsWrapper>
        <AvatarWrapper>
          <input
            ref={this.avatarInput}
            type="file"
            accept="image/png, image/jpg, image/jpeg"
            onChange={this.onChangeAvatar}
            style={{ display: 'none' }}
          />
          <Avatar
            src={avatar.src}
            alt={avatar.text}
            onClick={this.onAvatarClick}
          >
            <Camera fontSize="large" />
          </Avatar>
        </AvatarWrapper>
        <FormWrapper>
          {
            map(this.fields, ({ id, label }) => {
              const value = called[id] ? values[id] : me[id];
              const error = !!errors[id];
              const message = errors[id];
              const isUpdated = success === id;

              return (
                <TextField
                  key={id}
                  name={id}
                  value={value}
                  label={label}
                  error={error}
                  helperText={message}
                  variant="outlined"
                  margin="dense"
                  onChange={this.onChange}
                  onBlur={this.onSubmit}
                  onKeyDown={this.onKeyPressed}
                  InputProps={{
                    endAdornment: isUpdated && (
                      <InputCheckIconWrapper position="end">
                        <CheckIcon color="action" />
                      </InputCheckIconWrapper>
                    ),
                  }}
                  fullWidth
                  required
                />
              );
            })
          }
        </FormWrapper>
      </SettingsWrapper>
    );
  }
}

Settings.defaultProps = {
  me: {},
  updateUserResult: {},
  error: null,
};
Settings.propTypes = {
  me: PropTypes.shape(meProps),
  uploadAvatar: PropTypes.func.isRequired,
  updateUser: PropTypes.func.isRequired,
  updateUserResult: PropTypes.shape({
    field: PropTypes.string,
    value: PropTypes.string,
  }),
  error: PropTypes.instanceOf(Error),
};

export default Settings;
