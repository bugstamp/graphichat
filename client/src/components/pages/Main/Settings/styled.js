import styled, { keyframes } from 'styled-components';
import { position } from 'polished';
import { fadeOut } from 'react-animations';

import MaterialAvatar from '@material-ui/core/Avatar';
import CameraIcon from '@material-ui/icons/CameraAlt';
import InputAdornment from '@material-ui/core/InputAdornment';

import { getStyledProps, getSpacing } from '../../../../styles';

const fadeOutAnimation = keyframes`${fadeOut}`;

const SettingsWrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('xs');

    return `
      ${smDown} {
        flex-flow: column;
        justify-content: flex-start;
      }
    `;
  }}
`;

const InputCheckIconWrapper = styled(InputAdornment)`
  && {
    animation: ${fadeOutAnimation};
    animation-duration: .5s;
    animation-delay: 5s;
    animation-fill-mode: forwards;
  }
`;

const AvatarWrapper = styled.div`
  position: relative;
  display: flex;
`;

const Avatar = styled(MaterialAvatar)`
  && {
    width: 172px;
    height: 172px;
    margin: 0 auto;
    position: relative;
    color: ${({ src }) => (src && '#fff')};
    cursor: pointer;

    ${({ src }) => src && `
      > svg {
        opacity: 0;
      }

      &:hover {
        > svg {
          opacity: .6;
        }
      }
    `}
  }
`;

const Camera = styled(CameraIcon)`
  && {
    ${position('absolute', '50%', null, null, '50%')};
    transform: translate(-50%, -50%);
    transition: all .3s ease;
  }
`;

const FormWrapper = styled.div`
  width: 100%;
  padding: ${getSpacing(1)};
`;

export {
  SettingsWrapper,
  AvatarWrapper,
  Avatar,
  Camera,
  FormWrapper,
  InputCheckIconWrapper,
};
