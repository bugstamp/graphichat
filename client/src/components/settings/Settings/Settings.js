import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/react-hooks';

import SettingsAvatar from './SettingsAvatar';
import SettingsForm from './SettingsForm';

import {
  SettingsWrapper,
} from './styled';

import gql from '../../../gql';
import { userUpdate, uploadAvatarUpdate } from '../../../gql/updates/user';
import useAvatar from '../../hooks/useAvatar';

const {
  GET_ME,
  UPLOAD_AVATAR,
  UPDATE_USER,
} = gql;

const Settings = (props) => {
  const { mode, setMode, setLoading } = props;

  const { data: { me = {} } = {} } = useQuery(GET_ME);
  const [uploadAvatar, { loading: uploading }] = useMutation(UPLOAD_AVATAR, {
    update: uploadAvatarUpdate,
  });
  const [updateUser, updateUserResult] = useMutation(UPDATE_USER, {
    update: userUpdate,
  });
  const { loading: updating, called, error } = updateUserResult;
  const loading = uploading || updating;
  const avatar = useAvatar(me, 'md');

  useEffect(() => {
    if (called && !updating && !error) {
      setMode('read');
    }
  }, [called, updating, error, setMode]);

  useEffect(() => {
    setLoading(loading);
  }, [loading, setLoading]);

  return (
    <SettingsWrapper>
      <SettingsAvatar
        avatar={avatar}
        uploadAvatar={uploadAvatar}
      />
      <SettingsForm
        mode={mode}
        user={me}
        updateUser={updateUser}
        updateUserResult={updateUserResult}
      />
    </SettingsWrapper>
  );
};

Settings.propTypes = {
  mode: PropTypes.oneOf(['read', 'edit']).isRequired,
  setMode: PropTypes.func.isRequired,
  setLoading: PropTypes.func.isRequired,
};

export default memo(Settings);
