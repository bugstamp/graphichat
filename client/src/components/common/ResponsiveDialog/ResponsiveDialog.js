import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';
import Button from '@material-ui/core/Button';
import Slide from '@material-ui/core/Slide';
import Fade from '@material-ui/core/Fade';
import useMediaQuery from '@material-ui/core/useMediaQuery';

import TopProgressLine from '../TopProgressLine';

import { getStyledProps } from '../../../styles';

const ResponsiveDialogTitle = styled(DialogTitle)`
  && {
  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${smDown} {
        padding: ${spacing(1)}px ${spacing(2)}px;
      }
    `;
  }}
  }
`;

const ResponsiveDialogContent = styled(DialogContent)`
  && {
    flex: 1 auto;
    display: flex;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const spacing = getStyledProps('theme.spacing')(props);
    const smDown = breakpoints.down('sm');

    return `
      ${smDown} {
        padding: ${spacing(1)}px ${spacing(2)}px;
      }
    `;
  }}
  }
`;

const ResponsiveDialog = ({
  open,
  toggle,
  title,
  subTitle,
  loading,
  children,
  ...rest
}) => {
  const fullScreen = useMediaQuery(theme => theme.breakpoints.down('sm'));
  const TransitionComponent = fullScreen ? Slide : Fade;
  const TransitionProps = fullScreen
    ? { direction: 'up' }
    : {};
  const PaperProps = {
    elevation: fullScreen ? 0 : 1,
  };

  return (
    <Dialog
      open={open}
      onClose={toggle}
      fullScreen={fullScreen}
      PaperProps={PaperProps}
      TransitionComponent={TransitionComponent}
      TransitionProps={TransitionProps}
      hideBackdrop={fullScreen}
      scroll="paper"
      disableEscapeKeyDown
      {...rest}
    >
      <TopProgressLine loading={loading} />
      <ResponsiveDialogTitle disableTypography>
        <Typography variant="h6">
          {title}
        </Typography>
        <If condition={subTitle}>
          <Typography variant="subtitle2">
            {subTitle}
          </Typography>
        </If>
      </ResponsiveDialogTitle>
      <ResponsiveDialogContent dividers>
        {children}
      </ResponsiveDialogContent>
      <DialogActions>
        <Button onClick={toggle} color="primary">Close</Button>
      </DialogActions>
    </Dialog>
  );
};

ResponsiveDialog.defaultProps = {
  loading: false,
  subTitle: '',
};
ResponsiveDialog.propTypes = {
  loading: PropTypes.bool,
  open: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node, PropTypes.element]).isRequired,
};

export default ResponsiveDialog;
