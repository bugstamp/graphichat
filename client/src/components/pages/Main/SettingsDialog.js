import React, { memo } from 'react';
import PropTypes from 'prop-types';

import ResponsiveDialog from '../../common/ResponsiveDialog/ResponsiveDialog';
import Settings from './Settings';

import SettingsDialogContainer from '../../containers/SettingsDialogContainer';

const SettingsDialog = ({ open, toggle }) => {
  return (
    <SettingsDialogContainer>
      {({
        getMe: {
          data = {},
        },
        uploadAvatar: {
          mutation: uploadAvatar,
          result: {
            loading: avatarUploading,
          },
        },
        updateUser: {
          mutation: updateUser,
          result: {
            loading: updating,
            error,
          },
        },
      }) => {
        const { me } = data;
        const loading = avatarUploading || updating;

        return (
          <ResponsiveDialog
            open={open}
            toggle={toggle}
            loading={loading}
            title="Settings"
          >
            <Settings
              me={me}
              error={error}
              updateUser={updateUser}
              uploadAvatar={uploadAvatar}
            />
          </ResponsiveDialog>
        );
      }}
    </SettingsDialogContainer>
  );
};

SettingsDialog.defaultProps = {
  open: false,
};
SettingsDialog.propTypes = {
  open: PropTypes.bool,
  toggle: PropTypes.func.isRequired,
};

export default memo(SettingsDialog);
