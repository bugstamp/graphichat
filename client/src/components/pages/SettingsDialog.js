import React, { PureComponent, createRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
// import {} from 'polished';

import DialogTitle from '@material-ui/core/DialogTitle';
import DialogActions from '@material-ui/core/DialogActions';
import MaterialDialogContent from '@material-ui/core/DialogContent';
import MaterialAvatar from '@material-ui/core/Avatar';
import CameraIcon from '@material-ui/icons/CameraAlt';
import Button from '@material-ui/core/Button';

import ResponsiveDialog from '../common/ResponsiveDialog';
import TopProgressLine from '../common/TopProgressLine';

const DialogContent = styled(MaterialDialogContent)`
  && {
    width: 100%;
  }
`;

const AvatarWrapper = styled.div`
  width: 50%;
  max-width: 250px;
  height: auto;
  margin: 0 auto;
  position: relative;
`;

const Avatar = styled(MaterialAvatar)`
  && {
    width: 100%;
    height: 100%;
    cursor: pointer;
  }
`;

class SettingsDialog extends PureComponent {
  avatarInput = createRef();

  onChangeAvatar = async ({
    target: {
      files: [file],
      validity: { valid },
    },
  }) => {
    const { uploadAvatar } = this.props;

    if (valid) {
      uploadAvatar({ variables: { file } });
    }
  }

  onAvatarClick = () => {
    if (this.avatarInput) {
      this.avatarInput.current.click();
    }
  }

  render() {
    const {
      open,
      toggle,
      avatar,
      avatarUploading,
    } = this.props;

    return (
      <ResponsiveDialog
        open={open}
        toggle={toggle}
        maxWidth="sm"
        fullWidth
      >
        <TopProgressLine loading={avatarUploading} />
        <DialogTitle>Profile Settings</DialogTitle>
        <DialogContent dividers>
          <AvatarWrapper>
            <input
              ref={this.avatarInput}
              type="file"
              accept="image/png, image/jpg, image/jpeg"
              onChange={this.onChangeAvatar}
              style={{ display: 'none' }}
            />
            <Avatar
              src={avatar.src}
              alt={avatar.text}
              onClick={this.onAvatarClick}
            >
              <If condition={!avatar.src}>
                <CameraIcon fontSize="large" />
              </If>
            </Avatar>
          </AvatarWrapper>
        </DialogContent>
        <DialogActions>
          <Button onClick={toggle} color="primary">Close</Button>
        </DialogActions>
      </ResponsiveDialog>
    );
  }
}

SettingsDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired,
  avatar: PropTypes.shape({
    src: PropTypes.string,
    text: PropTypes.string,
  }).isRequired,
  avatarUploading: PropTypes.bool.isRequired,
  uploadAvatar: PropTypes.func.isRequired,
};

export default SettingsDialog;
