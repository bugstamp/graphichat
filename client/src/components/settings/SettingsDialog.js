import React, { memo, useState, useCallback } from 'react';
import PropTypes from 'prop-types';

import Typography from '@material-ui/core/Typography';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';
import Button from '@material-ui/core/Button';

import ResponsiveDialog from '../common/ResponsiveDialog';
import TopProgressLine from '../common/TopProgressLine';
import Settings from './Settings';

const SettingsDialog = (props) => {
  const { open, toggle } = props;
  const [mode, setMode] = useState('read');
  const [loading, setLoading] = useState(false);
  const buttonText = mode === 'edit' ? 'Cancel' : 'Edit';

  const handleToggleMode = useCallback(() => {
    setMode(mode === 'read' ? 'edit' : 'read');
  }, [mode]);

  return (
    <ResponsiveDialog
      open={open}
      toggle={toggle}
    >
      <TopProgressLine loading={loading} />
      <DialogTitle disableTypography>
        <Typography variant="h6">Settings</Typography>
      </DialogTitle>
      <DialogContent dividers>
        <Settings
          mode={mode}
          setMode={setMode}
          setLoading={setLoading}
        />
      </DialogContent>
      <DialogActions>
        <Button
          onClick={toggle}
          color="primary"
          type="button"
        >
          Close
        </Button>
        <Button
          onClick={handleToggleMode}
          color="primary"
          type="button"
          variant="contained"
        >
          {buttonText}
        </Button>
      </DialogActions>
    </ResponsiveDialog>
  );
};

SettingsDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired,
};

export default memo(SettingsDialog);
