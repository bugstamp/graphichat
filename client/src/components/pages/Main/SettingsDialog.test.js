import React from 'react';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import { shallow } from 'enzyme';
import { GraphQLError } from 'graphql';

import {
  getMeMock,
  uploadAvatarMock,
  updateUserMock,
} from '../../../__mocks__/mockedQueries';
import {
  me,
  updateUserResult,
} from '../../../__mocks__/mockedQueryData';
import { mountMockedProvider } from '../../../__mocks__/mockedProvider';
import { BadInputError } from '../../../__mocks__/mockedErrors';

import ResponsiveDialog from '../../common/ResponsiveDialog/ResponsiveDialog';
import SettingsDialog from './SettingsDialog';
import Settings from './Settings/Settings';

const defaultMocks = [
  getMeMock,
  uploadAvatarMock,
  updateUserMock,
];

describe('test SettingsDialog', () => {
  let open = false;
  const mockToggleFn = jest.fn(() => { open = !open; });
  const defaultProps = {
    open: false,
    toggle: mockToggleFn,
  };

  beforeEach(() => { open = false; });

  test('snapshot render', () => {
    const wrapper = shallow(<SettingsDialog />);

    expect(wrapper).toMatchSnapshot();
  });
  test('mount render', async () => {
    const wrapper = mountMockedProvider(<SettingsDialog {...defaultProps} />, defaultMocks);

    expect(wrapper.find(ResponsiveDialog)).toBeTruthy();
    expect(wrapper.find(Settings)).toBeTruthy();
  });
  test('getMe query', async () => {
    const wrapper = mountMockedProvider(<SettingsDialog {...defaultProps} open />, defaultMocks);

    await act(async () => {
      await wait();
      wrapper.update();
      expect(wrapper.find(Settings).prop('me')).toEqual(me);
    });
  });
  test('updateUser mutation', async () => {
    const wrapper = mountMockedProvider(<SettingsDialog {...defaultProps} open />, defaultMocks);
    expect(wrapper.find(ResponsiveDialog).prop('loading')).toBe(false);

    await act(async () => {
      await wait();
      wrapper.find('input[name="username"]').simulate('blur', { target: { name: 'username', value: 'user' } });
      expect(wrapper.find(ResponsiveDialog).prop('loading')).toEqual(true);
    });
    await act(async () => {
      await wait();
      wrapper.update();
      expect(wrapper.find(ResponsiveDialog).prop('loading')).toEqual(false);
      expect(wrapper.find(Settings).prop('updateUserResult')).toEqual(updateUserResult);
    });
  });
  test('updateUser mutation error', async () => {
    const error = new BadInputError({
      message: 'invalidField',
      data: {
        invalidField: 'username',
      },
    });
    const wrapper = mountMockedProvider(
      <SettingsDialog {...defaultProps} open />,
      [getMeMock, uploadAvatarMock, {
        ...updateUserMock,
        result: {
          errors: [error],
        },
      }],
    );

    await act(async () => {
      await wait();
      wrapper.find('input[name="username"]').simulate('blur', { target: { name: 'username', value: 'user' } });
    });
    await act(async () => {
      await wait();
      wrapper.update();
      expect(wrapper.find(Settings).prop('error')).toBeDefined();
      expect(wrapper.find(Settings).instance().state).toHaveProperty('errors.username');
    });
  });
  test('uploadAvatar mutation', async () => {
    const file = {
      filename: '',
      mimetype: '',
      encoding: '',
    };
    const wrapper = mountMockedProvider(<SettingsDialog {...defaultProps} open />, defaultMocks);
    expect(wrapper.find(ResponsiveDialog).prop('loading')).toBe(false);

    await act(async () => {
      await wait();
      wrapper.find('input[type="file"]').simulate('change', {
        target: {
          files: [file],
          validity: { valid: true },
        },
      });
      expect(wrapper.find(ResponsiveDialog).prop('loading')).toEqual(true);
    });
    await act(async () => {
      await wait();
      wrapper.update();
      expect(wrapper.find(ResponsiveDialog).prop('loading')).toEqual(false);
    });
  });
  test('pass open prop | call toggle func | open settings dialog', async () => {
    let wrapper = mountMockedProvider((
      <SettingsDialog {...defaultProps} open={open} />
    ), defaultMocks);
    expect(wrapper.find(ResponsiveDialog).prop('open')).toBe(false);
    expect(wrapper.find(ResponsiveDialog).prop('toggle')).toEqual(mockToggleFn);
    expect(wrapper.find(Settings)).toHaveLength(0);

    await act(async () => {
      await wait();
      wrapper.update();
      mockToggleFn();
      wrapper = mountMockedProvider(<SettingsDialog {...defaultProps} open={open} />, defaultMocks);
      expect(wrapper.find(ResponsiveDialog).prop('open')).toBe(true);
      expect(wrapper.find(Settings)).toHaveLength(1);
    });
  });
});
