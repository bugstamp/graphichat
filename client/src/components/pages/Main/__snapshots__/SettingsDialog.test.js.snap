// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SettingsDialog snapshot render 1`] = `
<Adopt(Unknown)(Unknown)(Unknown)>
  <Component />
</Adopt(Unknown)(Unknown)(Unknown)>
`;
